pkg/cri/server/sandbox_portforward_windows.go

```diff
@@ -17,61 +17,97 @@
 package server
 
 import (
-	"bytes"
 	"context"
 	"fmt"
 	"io"
+	"net"
+	"time"
 
-	"k8s.io/utils/exec"
-
-	sandboxstore "github.com/containerd/containerd/v2/pkg/cri/store/sandbox"
-	cioutil "github.com/containerd/containerd/v2/pkg/ioutil"
+	"github.com/containerd/log"
 )
 
 func (c *criService) portForward(ctx context.Context, id string, port int32, stream io.ReadWriter) error {
-	stdout := cioutil.NewNopWriteCloser(stream)
-	stderrBuffer := new(bytes.Buffer)
-	stderr := cioutil.NewNopWriteCloser(stderrBuffer)
-	// localhost is resolved to 127.0.0.1 in ipv4, and ::1 in ipv6.
-	// Explicitly using ipv4 IP address in here to avoid flakiness.
-	cmd := []string{"wincat.exe", "127.0.0.1", fmt.Sprint(port)}
-	err := c.execInSandbox(ctx, id, cmd, stream, stdout, stderr)
+	sandbox, err := c.sandboxStore.Get(id)
 	if err != nil {
-		return fmt.Errorf("failed to execute port forward in sandbox: %s: %w", stderrBuffer.String(), err)
+		return fmt.Errorf("failed to find sandbox %q in store: %w", id, err)
 	}
-	return nil
-}
 
-func (c *criService) execInSandbox(ctx context.Context, sandboxID string, cmd []string, stdin io.Reader, stdout, stderr io.WriteCloser) error {
-	// Get sandbox from our sandbox store.
-	sb, err := c.sandboxStore.Get(sandboxID)
-	if err != nil {
-		return fmt.Errorf("failed to find sandbox %q in store: %w", sandboxID, err)
+	var podIP string
+	if !hostNetwork(sandbox.Config) {
+		// get ip address of the sandbox
+		podIP, _, err = c.getIPs(sandbox)
+		if err != nil {
+			return fmt.Errorf("failed to get sandbox ip: %w", err)
+		}
+	} else {
+		// port forwarding is not supported for host process containers
+		return fmt.Errorf("port forwarding not supported for host process containers")
 	}
 
-	// Check the sandbox state
-	state := sb.Status.Get().State
-	if state != sandboxstore.StateReady {
-		return fmt.Errorf("sandbox is in %s state", fmt.Sprint(state))
-	}
+	err = func() error {
+		var conn net.Conn
+		conn, err = net.Dial("tcp4", fmt.Sprintf("%s:%d", podIP, port))
```

Comment:
We can determine podIP is either IPv4 or IPv6 that would make this a little cleaner?

Reply:
`func checkIPAddressType(ipAddress string) (string, error) {
	// ParseIP returns nil if the address is not a valid IPv4 or IPv6 address.
	ip := net.ParseIP(ipAddress)
	if ip == nil {
		return "", fmt.Errorf("Invalid IP address: %s", ipAddress)
	}

	// Check if it's an IPv4 address.
	if ip.To4() != nil {
		return "IPv4", nil
	}

	// Check if it's an IPv6 address.
	if ip.To16() != nil {
		return "IPv6", nil
	}

	return "", fmt.Errorf("Unknown IP address type for: %s", ipAddress)
}`

Reply:
I commented above around the same bits, but this seems to be from the linux implementation of this. The workaround they have was for localhost connections (because they join the namespace) and shouldn't be relevant here, but yea we should figure out what kind of ip it is so we can format correctly for dial.

Reply:
https://github.com/containerd/containerd/blob/4f3a026bceb629f5d912e07bbede123b8593d340/vendor/k8s.io/utils/net/ipfamily.go#L166 There is k8s.io util function that i could use. Does this look ok?